packages/amplify_core/lib/src/types/common/amplify_auth_provider.dart

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:aws_signature_v4/aws_signature_v4.dart';
+
+/// An identifier to use as a key in an [AmplifyAuthProviderRepository] so that
+/// a retrieved auth provider can be typed more accurately.
+class AmplifyAuthProviderToken<T extends AmplifyAuthProvider> extends Token<T> {
+  const AmplifyAuthProviderToken();
+}
+
+abstract class AuthProviderOptions {
+  const AuthProviderOptions();
+}
+
+/// Options required by IAM to sign any given request at runtime.
+class IamAuthProviderOptions extends AuthProviderOptions {
+  final String region;
+  final AWSService service;
+
+  const IamAuthProviderOptions({required this.region, required this.service});
+}
+
+abstract class AmplifyAuthProvider {
+  Future<AWSBaseHttpRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant AuthProviderOptions? options,
```

Comment:
Good to know about `covariant`, it should help us to avoid type casting required on consumer side like this?

```dart
    final result = await Amplify.Auth.fetchAuthSession(
      options: CognitoSessionOptions(getAWSCredentials: true),
    );
    String identityId = (result as CognitoAuthSession).identityId!;
```

Reply:
Maybe I'm missing this, but don't think it helps here and casting would still be needed as I think it can only be used for parameters, not return types.

Separately, I think we will be able to define an auth provider that can provide identity IDs so casting would not be needed outside auth plugin for this use case.